s2a/src/main/java/io/grpc/s2a/handshaker/S2AStub.java

```diff
@@ -0,0 +1,221 @@
+   *
+   * @param req the {@code SessionReq} message to be sent to the S2A server.
+   * @return the {@code SessionResp} message received from the S2A server.
+   * @throws ConnectionClosedException if {@code reader} or {@code writer} calls their {@code
+   *     onCompleted} method.
+   * @throws IOException if an unexpected response is received, or if the {@code reader} or {@code
+   *     writer} calls their {@code onError} method.
+   */
+  public SessionResp send(SessionReq req) throws IOException, InterruptedException {
+    if (doneWriting && doneReading) {
+      logger.log(Level.INFO, "Stream to the S2A is closed.");
+      throw new ConnectionClosedException("Stream to the S2A is closed.");
+    }
+    createWriterIfNull();
+    if (!responses.isEmpty()) {
+      IOException exception = null;
+      SessionResp resp = null;
+      try {
+        resp = responses.take().getResultOrThrow();
+      } catch (IOException e) {
+        exception = e;
+      }
+      responses.clear();
+      if (exception != null) {
+        throw new IOException(
+            "Received an unexpected response from a host at the S2A's address. The S2A might be"
+                + " unavailable."
+                + exception.getMessage());
+      }
+      return resp;
+    }
+    try {
+      writer.onNext(req);
+    } catch (RuntimeException e) {
+      writer.onError(e);
+      responses.offer(Result.createWithThrowable(e));
+    }
+    try {
+      return responses.take().getResultOrThrow();
+    } catch (ConnectionClosedException e) {
+      // A ConnectionClosedException is thrown by getResultOrThrow when reader calls its
+      // onCompleted method. The close method is called to also close the writer, and then the
+      // ConnectionClosedException is re-thrown in order to indicate to the caller that send
+      // should not be called again.
+      close();
+      throw e;
+    }
+  }
+
+  @Override
+  public void close() {
+    if (doneWriting && doneReading) {
+      return;
+    }
+    verify(!doneWriting);
+    doneReading = true;
+    doneWriting = true;
+    if (writer != null) {
+      writer.onCompleted();
+    }
+  }
+
+  /** Create a new writer if the writer is null. */
+  private void createWriterIfNull() {
+    if (writer == null) {
+      writer =
+          serviceStub
+              .withWaitForReady()
+              .withDeadlineAfter(HANDSHAKE_RPC_DEADLINE_SECS, SECONDS)
+              .setUpSession(reader);
+    }
+  }
+
+  private class Reader implements StreamObserver<SessionResp> {
+    /**
+     * Places a {@code SessionResp} message in the {@code responses} queue, or an {@code
+     * IOException} if reading is complete.
+     *
+     * @param resp the {@code SessionResp} message received from the S2A handshaker module.
+     */
+    @Override
+    public void onNext(SessionResp resp) {
+      verify(!doneReading);
+      responses.offer(Result.createWithResponse(resp));
```

Comment:
All the `offer()`s scare me. The queue is sort of randomly sized (not just 2), which also doesn't feel me with confidence. It'd be much better to use `add()` which throws an exception if the queue is full instead of silently losing data.

Reply:
Done in https://github.com/grpc/grpc-java/pull/11540/commits/d13bd88b66d4bb24d1879b3afe1e961bf212fb96